FEATURE: New `Video` API

Ashp116 opened the pull request:
# Description

This PR introduces a new `Video` API that streamlines video processing and rendering workflows. It addresses both issues #1923 and #1929 by enabling more flexible backend support and improved audio-video synchronization.

With this update, the video processing function now supports multiple backends, including `PyAV` and `OpenCV`. Notably, `PyAV` is the only backend currently supporting audio rendering, which significantly improves output quality.

This PR requires the optional dependency [`pyAV`](https://pypi.org/project/av/) for the video rendering backend. 

Tags:
Fixes #1923 
Fixes #1929

## Type of change

* [x] Bug fix (non-breaking change which fixes an issue)

## How has this change been tested, please provide a testcase or example of how you tested the change?

Please refer to #1923 and #1929

## Any specific deployment considerations

Ensure that [`pyAV`](https://pypi.org/project/av/) is installed in the environment to test `pyAV` backend.

## Docs

* [ ] Docs updated? What were the changes

Ashp116 committed:
ADD: Added audio stream for process_video

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
REMOVE: Removed ffprobe

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

SkalskiP commented:
Hi @Ashp116 👋🏻 Another great idea! Video processing is probably the oldest part of `supervision`, written over two years ago, and I’ve been wanting to update its API for a while. Would you be open to not only adding audio support but also helping me with the update?

Ashp116 commented:
Hi @SkalskiP, yea, I would like to help update the API. I was thinking of changing how videos are written in `process_video`. The original compression is lost when annotations are added and the file is written to a `target_path`. But yea, I would like to help out with the update.

SkalskiP commented:
Hi @Ashp116 bardzo się cieszę, że chcesz mi pomoc! Let's goooo! 🔥 🔥 🔥

I want the functionalities currently found in `supervision.utils.video` to be reorganized around a new `Video` class. Importantly, all features previously available in the old API must still be supported in the new one. Ideally, the new API should be more consistent and expressive.

- get video info (works for files, RTSP, webcams)

    ```python
    import supervision as sv
     
    # static video
    sv.Video("source.mp4").info
    
    # video stream
    sv.Video("rtsp://...").info

    # webcam
    sv.Video(0).info
    ```

- simple frame iteration (object is iterable)

    ```python
    import supervision as sv
    
    video = sv.Video("source.mp4")
    for frame in video:
        ...
    ```

- advanced frame iteration (stride, sub-clip, on-the-fly resize)

    ```python
    import supervision as sv
    
    for frame in sv.Video("source.mp4").frames(stride=5, start=100, end=500, resolution_wh=(1280, 720)):
        ...
    ```

- process the video

    ```python
    import cv2
    import supervision as sv

    def blur(frame, i):
        return cv2.GaussianBlur(frame, (11, 11), 0)
    
    sv.Video("source.mp4").save(
        "blurred.mp4",
        callback=blur,
        show_progress=True
    )
    ```

- overwrite target video parameters

    ```python
    import supervision as sv
    
    sv.Video("source.mp4").save(
        "timelapse.mp4",
        fps=60,
        callback=lambda f, i: f,
        show_progress=True
    )
    ```

- complete manual control with explicit `VideoInfo`

    ```python
    from supervision import Video, VideoInfo
    
    source = Video("source.mp4")
    target_info = VideoInfo(width=800, height=800, fps=24)
    
    with src.sink("square.mp4", info=target_info) as sink:
        for f in src.frames():
            f = cv2.resize(f, target_info.resolution_wh)
            sink.write(f)
    ```

- multi-backend support decode/encode

    ```python
    import supervision as sv
    
    video = sv.Video("source.mkv", backend="pyav")
    
    video = sv.Video("source.mkv", backend="opencv")
    ```

    suggested minimal protocol

    ```python
    class Backend(Protocol):
        def open(self, path: str) -> Any: ...
        def info(self, handle: Any) -> VideoInfo: ...
    
        def read(self, handle: Any) -> tuple[bool, np.ndarray]: ...
        def grab(self, handle: Any) -> bool: ...
        def seek(self, handle: Any, frame_idx: int) -> None: ...
    
        def writer(self, path: str, info: VideoInfo, codec: str) -> Writer: ...
    
    class Writer(Protocol):
        def write(self, frame: np.ndarray) -> None: ...
        def close(self) -> None: ...
    ```

Ashp116 committed:
UPDATE: Added a new Video class with OpenCV writer and backend

Ashp116 committed:
Merge pull request #1 from Ashp116/update/video-core

UPDATE: Added a new Video class with OpenCV writer and backend

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
Precommit

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
Precommit

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Fixed incomplete write closing

Ashp116 committed:
ADD: Docstrings

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Allow for ffmpeg error passthrough

Ashp116 committed:
UPDATE: Writer and Backend abstract class

Ashp116 committed:
Precommit

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
Precommit

Ashp116 commented:
Hi @SkalskiP,

I’ve addressed most of the features you mentioned, but I have some thoughts on a few aspects of the implementation:

* **`.save` Functionality**
  How would you handle `.save` for a video feed coming from a webcam or an RTSP stream? Currently, I have it where only video files can be saved.

* **Writer and Backend Classes**
  This is just my personal opinion, but should these classes be moved to separate scripts/modules? If we add more writers and backends in the future, keeping everything inside the main video script might become cluttered.

* **“Complete manual control with explicit `VideoInfo`” Functionality**

  ```python
  from supervision import Video, VideoInfo

  source = Video("source.mp4")
  target_info = VideoInfo(width=800, height=800, fps=24)

  with src.sink("square.mp4", info=target_info) as sink:
      for f in src.frames():
          f = cv2.resize(f, target_info.resolution_wh)
          sink.write(f)
  ```

  I’m not fully clear on what this feature is intended to do. In this snippet, the `Video` instance `source` is created but never used afterward. Is `src` supposed to be `source`? Also, is the goal to create sinks for each backend? Could you please clarify the purpose and expected usage here?

Ashp116 committed:
UPDATE: Added manual control

Ashp116 committed:
ADD: Added docstrings

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Deprecate warning old Video API

Ashp116 committed:
FIX: Prototype resolution for #1687

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
FIX: Rounding type

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

SkalskiP commented on pyproject.toml:
```diff
@@ -46,7 +46,8 @@ dependencies = [
     "tqdm>=4.62.3",
-    "opencv-python>=4.5.5.64"
+    "opencv-python>=4.5.5.64",
+    "imageio-ffmpeg (>=0.6.0,<0.7.0)"
```
Let's not use `imageio-ffmpeg` and use `PyAV` instead.

SkalskiP replied:
When adding `PyAV` as dependency, make sure to make it optional. We don't want to require everyone to install it.

SkalskiP commented on supervision/utils/video.py:
```diff
@@ -141,6 +673,7 @@ def _validate_and_setup_video(
     return video, start, end
 
 
+@DeprecationWarning
```
let's use `deprecated` from `supervision.utils.internal`. it is currently used in the codebase to mark several deprecations.

SkalskiP commented on supervision/utils/video.py:
```diff
@@ -192,6 +725,7 @@ def get_video_frames_generator(
     video.release()
 
 
+@DeprecationWarning
```
let's use `deprecated` from `supervision.utils.internal`. it is currently used in the codebase to mark several deprecations.

SkalskiP commented on supervision/utils/video.py:
```diff
@@ -117,6 +648,7 @@ def __exit__(self, exc_type, exc_value, exc_traceback):
         self.__writer.release()
 
 
+@DeprecationWarning
```
No need to mark private classes / functions as deprecated. We do it only for classes / functions in public supervision API.

SkalskiP commented on supervision/utils/video.py:
```diff
@@ -60,6 +74,523 @@ def resolution_wh(self) -> tuple[int, int]:
+        )
+
+
+@DeprecationWarning
```
let's use `deprecated` from `supervision.utils.internal`. it is currently used in the codebase to mark several deprecations.

SkalskiP replied:
Internals of this deprecated class should be now reimplemented using new `Video` API.

SkalskiP commented on supervision/utils/video.py:
```diff
@@ -141,6 +673,7 @@ def _validate_and_setup_video(
 
 
+@DeprecationWarning
 def get_video_frames_generator(
```
Internals of this deprecated function should be now reimplemented using new `Video` API.

Ashp116 committed:
UPDATE: Updated the file structure for the video API

Ashp116 committed:
UPDATE: Added frame generator within backend

Ashp116 committed:
UPDATE: Added .save codec support

Ashp116 committed:
UPDATE: Added default mp4v codec

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Trying to pass checks

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

Ashp116 committed:
UPDATE: Typings for optional params

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
Import check fix

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
Import check fix

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
Import check fix

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
Import check fix

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

Ashp116 committed:
Import check fix

Ashp116 committed:
Import check fix

Ashp116 committed:
Import check fix

Ashp116 committed:
RM: Remove video folder

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
ADD: Added video folder

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

Ashp116 committed:
UPDATE: Revert typings

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: sv imports

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

Ashp116 committed:
UPDATE: default imports using __init__

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
FIX: Fixed VideoInfo definition

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
FIX: Fixed futuer annotations

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Restructure video saves

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

Ashp116 committed:
UPDATE: Restructure manual writer

Ashp116 committed:
FORMAT: Formatting changes

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Restructured frame generator

Ashp116 committed:
FIX: Removed frames from BaseBackend

Ashp116 committed:
ADD: Added pyAV backend support

Ashp116 committed:
ADD: Added pyAV docstrings

Ashp116 committed:
ADD: Added pyAV audio support

Ashp116 committed:
UPDATE: Backend support needed for all writers

Ashp116 committed:
FIX: Formatting

Ashp116 committed:
ADD: Added backend utils and type safe support

Ashp116 committed:
UPDATE: Type safe syntax

Ashp116 committed:
FORMAT: Changed backend param

Ashp116 committed:
UPDATE: Added ctx for pyAV

Ashp116 committed:
UPDATE: Updated docstrings

Ashp116 committed:
UPDATE: Negate audio stream

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
FIX: Fixed av import error and formatting

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
FIX: Fixed av error message and formatting

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
FIX: Fixed issue with audio sync with fps

Ashp116 committed:
UPDATE: Finer audio mux

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Param config to render audio

Ashp116 committed:
Merge branch 'bug/process-video-audio' of https://github.com/Ashp116/supervision into bug/process-video-audio

Ashp116 committed:
UPDATE: Changed config to render audio to be isolated

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Updated deprecated warnings

Ashp116 committed:
ADD: Added unit tests from PR #1941

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

SkalskiP commented on pyproject.toml:
```diff
@@ -58,6 +58,7 @@ Documentation = "https://supervision.roboflow.com/latest/"
 metrics = [
     "pandas>=2.0.0",
 ]
+video = ["av (>=15.0.0,<16.0.0)"]
```
I’d rename it from `video` to `ffmpeg`. Also, do we need to set the maximum version at this point?

Ashp116 replied:
I don't think we need the maximum version. Poetry just generated it.

SkalskiP commented on supervision/utils/video.py:
```diff
@@ -192,6 +197,7 @@ def get_video_frames_generator(
     video.release()
 
 
+@deprecated("Use `sv.Video.save()` with a callback for processing and saving videos.")
```
We aim to keep the format of all deprecation messages consistent. Please use the message below and ensure the same formatting is applied to all deprecation messages.

```
@deprecated(
    "`process_video` is deprecated and will be removed in "
    "`supervision-0.32.0`. Use `Video().save` instead."
)
```

SkalskiP commented on supervision/video/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from supervision.video.backend.base import BaseBackend, BaseWriter
+from supervision.video.core import Video
+from supervision.video.utils import SOURCE_TYPE, VideoInfo
+
+__all__ = ["SOURCE_TYPE", "BaseBackend", "BaseWriter", "Video", "VideoInfo"]
```
I don't think we need this.

SkalskiP commented on supervision/video/backend/openCV.py:
```diff
@@ -0,0 +1,213 @@
+from __future__ import annotations
```
Let’s make all file names snake_case, `opencv.py` instead of `openCV.py`.

SkalskiP commented on supervision/video/utils.py:
```diff
@@ -0,0 +1,104 @@
+    source_type: SOURCE_TYPE | None = None
+
+    @classmethod
+    def from_video_path(cls, video_path: str) -> VideoInfo:
+        """
+        Create a VideoInfo instance from a video file.
+
+        Args:
+            video_path (str): Path to the video file.
+
+        Returns:
+            VideoInfo: Metadata including width, height, FPS, and total frames.
+
+        Raises:
+            ValueError: If the video cannot be opened or has invalid properties.
+        """
+        video = cv2.VideoCapture(video_path)
+        if not video.isOpened():
+            raise ValueError(f"Could not open video at {video_path}")
+
+        try:
+            width = int(video.get(cv2.CAP_PROP_FRAME_WIDTH))
+            height = int(video.get(cv2.CAP_PROP_FRAME_HEIGHT))
+            if width <= 0 or height <= 0:
+                raise ValueError(f"Invalid video dimensions: {width}x{height}")
+
+            fps = video.get(cv2.CAP_PROP_FPS)
+            if fps <= 0:
+                fps = 30  # Default to 30fps if invalid
+            fps = round(fps)
+
+            total_frames = int(video.get(cv2.CAP_PROP_FRAME_COUNT))
+            if total_frames < 0:
+                total_frames = None  # Some formats may not report frame count
+        finally:
+            video.release()
+
+        return VideoInfo(width, height, fps, total_frames)
```
Shouldn't this implementation be backend-specific?

Ashp116 commented on supervision/video/backend/base.py:
```diff
@@ -0,0 +1,77 @@
+        frame_size: tuple[int, int],
+        codec: str | None = None,
+        backend: BaseBackend | None = None,
+        render_audio: bool = False,
```
I think we do. `pyAV`'s default compression codec is `h264`, which is way better than `opencv`'s `mp4v`. So if a user wants just to render the video frames and not the audio using `pyAV`, this allows for that. Although, I think we should set `render_audio` to `None` and the default to `True`.

Ashp116 committed:
FIX: Updated the pyproject.toml's extra requirements for ffmpeg

Ashp116 committed:
FIX: Fixes of formatting from code review

Ashp116 committed:
UPDATE: Removed video info from path as it is dependant on the backend

Ashp116 committed:
UPDATE: Updated backend to Enum

Ashp116 committed:
FIX: Reverted video utils

Ashp116 committed:
FIX: Fixed backend bug issues

Ashp116 committed:
UPDATE: Updated render_audio param default value config

Ashp116 committed:
BUG: Fixed pyav not resolving webcam source path

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Updated file names

Ashp116 committed:
Precommit error fix

Ashp116 committed:
FIX: render_audio opencv param fix

Ashp116 committed:
UPDATE: Updated docstrings

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Updated docstrings

Ashp116 committed:
UPDATE: Added .show()

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 commented:
Hi @SkalskiP,

Thank you for reviewing the PR. I have addressed all the comments from the review. Could you please take a look at the following points?

* **`render_audio` parameter:**
  You mentioned the need for this parameter. I agree it’s necessary. Here is my [response](https://github.com/roboflow/supervision/pull/1924#discussion_r2267654159):

> I think we do. `pyAV`’s default compression codec is `h264`, which produces much better quality than OpenCV’s `mp4v`. If a user wants to render only the video frames without audio using `pyAV`, this parameter allows that. I also suggest setting `render_audio` to `None` with the default as `True`.
Overall, `ffmpeg`’s default compression leads to better video outputs.

* **`.show()` feature:**
  I think this is a useful addition, and I have implemented it. However, there are some issues: currently, `cv2.imshow` is used to render the frame with a wait time of 1 ms. This causes the display to not match the correct FPS for a given video source. There are ways to address this, but they would require adding a dependency. Could you share your thoughts on this implementation approach?

* **`pyAV` webcam bug:**
  In my previous review, I completely missed webcam support for the `pyAV` backend. I’ve added this in the current PR. Using a webcam with `pyAV` requires a different code path. Could you help test this feature on other devices? I’ve verified it works on a Windows machine.

Please let me know if you have any feedback or suggestions. Thanks

Ashp116 committed:
UPDATE: Add support for IPython display

Ashp116 committed:
UPDATE: Added support for headless machines and notebook for sv.Video().show()

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Updated error msg for IPython

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
Precommit

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Fixed av module install

Ashp116 committed:
UPDATE: Revert av error

Ashp116 committed:
UPDATE: updated av module getter

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Updated .show() with more configuration params

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Updated IPython import

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
BUG: Frame iteration fix

Ashp116 committed:
UPDATE: Updated audio stream to use atempo reflecting changes in fps

Ashp116 committed:
UPDATE: Updated docstrings

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Changed backend type class and added ref to root

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
BUG: Appending fixes for VideoBackend error

Ashp116 committed:
FIX: Merge conflicts

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

Ashp116 committed:
UPDATE: Decompose playback speed into valid atempo chain

pre-commit-ci[bot] committed:
fix(pre_commit): 🎨 auto format pre-commit hooks

ryashry reviewed (changes requested):
Improve Hight quality video web

Ashp116 commented:
Hello @SkalskiP, 

I been a while, I hope things are going great! I was wondering if you can review this pr at its current state so I can make changes and add-in couple of other features to this pr?
